### PR TITLE
Added testcase for :Share backups and cluster with user

### PR DESCRIPTION
**Pull request by snigdha-px**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** (optional)
Closes #
https://portworx.atlassian.net/browse/PA-324
**Special notes for your reviewer**:
Expected behaviour : https://portworx.atlassian.net/browse/PB-3486
Duplicate backup is a UI testing . We need to check that duplicate button is not there . Create backup will go through as it does not have any reference to the backup shared
Aetos link : http://aetos.pwx.purestorage.com/resultSet/testSetID/77117
http://aetos.pwx.purestorage.com/resultSet/testSetID/77150

### CODE REVIEW

File: tests/backup/backup_test.go [hunk 1]
```diff
@@ -1303,6 +1325,131 @@ var _ = Describe("{BackupAlternatingBetweenLockedAndUnlockedBuckets}", func() {
...
+			providers := getProviders()
+			for _, provider := range providers {
+				bucketName := fmt.Sprintf("%s-%s", provider, bucketNames[0])
+				cloudCredName = fmt.Sprintf("%s-%s", "cloudcred", provider)
+				bkpLocationName = fmt.Sprintf("%s-%s-bl", provider, bucketNames[0])
+				cloudCredUID = uuid.New()
+				backupLocationUID = uuid.New()
+				backupLocationMap[backupLocationUID] = bkpLocationName
+				CreateCloudCredential(provider, cloudCredName, cloudCredUID, orgID)
+				time.Sleep(time.Minute * 4)
```
Comment (lsrinivas-pure) [inline tests/backup/backup_test.go:1393]: Add a to-do here to remove this wait with validation. also make sure there is a ticket for that task

File: tests/backup/backup_test.go [hunk 2]
```diff
@@ -1303,6 +1325,131 @@ var _ = Describe("{BackupAlternatingBetweenLockedAndUnlockedBuckets}", func() {
...
+		log.FailOnError(err, "Fetching px-central-admin ctx")
+		opts := make(map[string]bool)
+		opts[SkipClusterScopedObjects] = true
+		ValidateAndDestroy(contexts, opts)
+		log.Infof(" Deleting backup created by px-central-admin")
+		backupDriver := Inst().Backup
+		backupUID, err := backupDriver.GetBackupUID(ctx, backupName, orgID)
+		dash.VerifySafely(err, nil, "Getting backup UID")
+		DeleteBackup(backupName, backupUID, orgID, ctx)
+		time.Sleep(time.Minute * 3)
```
Comment (lsrinivas-pure) [inline tests/backup/backup_test.go:1437]: same here

File: tests/backup/backup_test.go (hunk 1 above)
Comment (snigdha-px) [reply tests/backup/backup_test.go:1393]: sure

File: tests/backup/backup_test.go (hunk 2 above)
Comment (snigdha-px) [reply tests/backup/backup_test.go:1437]: sure

File: tests/backup/backup_test.go [hunk 3]
```diff
@@ -1303,6 +1325,131 @@ var _ = Describe("{BackupAlternatingBetweenLockedAndUnlockedBuckets}", func() {
...
+		log.Infof(" Deleting backup created by user")
+		userBackupUID, err := backupDriver.GetBackupUID(ctxNonAdmin, userBackupName, orgID)
+		dash.VerifySafely(err, nil, "Getting backup UID of user")
+		DeleteBackup(userBackupName, userBackupUID, orgID, ctxNonAdmin)
+		time.Sleep(time.Minute * 3)
+		log.Infof("Cleaning up users")
+		err = backup.DeleteUser(userName)
+		log.FailOnError(err, "Error in deleting user")
+		log.Infof("Deleting registered clusters for non-admin context")
```
Comment (kshithijiyer-px) [inline tests/backup/backup_test.go:1449]: Can we use `DeleteCloudAccounts` here?

File: tests/backup/backup_test.go [hunk 4]
```diff
@@ -1303,6 +1325,131 @@ var _ = Describe("{BackupAlternatingBetweenLockedAndUnlockedBuckets}", func() {
...
+		Step("Create Users", func() {
+			log.InfoD("Creating users testuser with FullAccess")
+			userName = "testuser"
+			firstName := "FirstName"
+			lastName := "LastName"
+			email := "testuser@cnbu.com"
+			err := backup.AddUser(userName, firstName, lastName, email, "Password1")
+			log.FailOnError(err, "Failed to create user - %s", userName)
+		})
+		Step("Creating bucket,backup location and cloud setting", func() {
```
Comment (kshithijiyer-px) [inline tests/backup/backup_test.go:1381]: No buckets getting created here. Please fix log

File: tests/backup/backup_test.go [hunk 5]
```diff
@@ -1303,6 +1325,131 @@ var _ = Describe("{BackupAlternatingBetweenLockedAndUnlockedBuckets}", func() {
...
+		userBackupName    string
+		ctxNonAdmin       context.Context
+	)
+	backupLocationMap := make(map[string]string)
+	labelSelectors := make(map[string]string)
+	bkpNamespaces = make([]string, 0)
+	JustBeforeEach(func() {
+		StartTorpedoTest("ShareBackupsAndClusterWithUser",
+			"Share backup to user with full access and try to duplicate the backup from the shared user", nil, 82943)
+		log.InfoD("Deploy applications")
```
Comment (kshithijiyer-px) [inline tests/backup/backup_test.go:1352]: Can we use a better log message here?

File: tests/backup/backup_test.go [hunk 6]
```diff
@@ -1303,6 +1325,131 @@ var _ = Describe("{BackupAlternatingBetweenLockedAndUnlockedBuckets}", func() {
...
+		})
+	})
+	JustAfterEach(func() {
+		defer EndTorpedoTest()
+		ctx, err := backup.GetAdminCtxFromSecret()
+		log.FailOnError(err, "Fetching px-central-admin ctx")
+		opts := make(map[string]bool)
+		opts[SkipClusterScopedObjects] = true
+		ValidateAndDestroy(contexts, opts)
+		log.Infof(" Deleting backup created by px-central-admin")
```
Comment (kshithijiyer-px) [inline tests/backup/backup_test.go:1433]: Remove the space here?

File: tests/backup/backup_test.go [hunk 7]
```diff
@@ -1303,6 +1325,131 @@ var _ = Describe("{BackupAlternatingBetweenLockedAndUnlockedBuckets}", func() {
...
+		backupDriver := Inst().Backup
+		backupUID, err := backupDriver.GetBackupUID(ctx, backupName, orgID)
+		dash.VerifySafely(err, nil, "Getting backup UID")
+		DeleteBackup(backupName, backupUID, orgID, ctx)
+		time.Sleep(time.Minute * 3)
+		log.Infof(" Deleting backup created by user")
```
Comment (kshithijiyer-px) [inline tests/backup/backup_test.go:1440]: Remove the space here.

File: tests/backup/backup_test.go [hunk 8]
```diff
@@ -1303,6 +1325,131 @@ var _ = Describe("{BackupAlternatingBetweenLockedAndUnlockedBuckets}", func() {
...
+			err = CreateBackup(backupName, SourceClusterName, bkpLocationName, backupLocationUID, []string{bkpNamespaces[0]},
+				labelSelectors, orgID, clusterUid, "", "", "", "", ctx)
+			dash.VerifyFatal(err, nil, "Verifying backup creation")
+		})
+		Step("Share backup with  user having full access", func() {
+			log.InfoD("Share backup with  user having full access")
+			ShareBackup(backupName, nil, []string{userName}, FullAccess, ctx)
+		})
+		Step("Create  backup from the shared user with FullAccess", func() {
+			log.InfoD(" Validating if user with  FullAccess cannot duplicate backup shared but can create new backup")
```
Comment (kshithijiyer-px) [inline tests/backup/backup_test.go:1414]: Remove space here

File: tests/backup/backup_test.go (hunk 3 above)
Comment (snigdha-px) [reply tests/backup/backup_test.go:1449]: DeleteCloudAccounts deletes backup location and cloud cred also.
for user we do not have that, hence used individual method

Comment (kshithijiyer-px) [reply tests/backup/backup_test.go:1449]: Okay

File: tests/backup/backup_test.go (hunk 4 above)
Comment (snigdha-px) [reply tests/backup/backup_test.go:1381]: done!

File: tests/backup/backup_test.go (hunk 5 above)
Comment (snigdha-px) [reply tests/backup/backup_test.go:1352]: done

File: tests/backup/backup_test.go (hunk 6 above)
Comment (snigdha-px) [reply tests/backup/backup_test.go:1433]: done

File: tests/backup/backup_test.go (hunk 7 above)
Comment (snigdha-px) [reply tests/backup/backup_test.go:1440]: done

File: tests/backup/backup_test.go (hunk 8 above)
Comment (snigdha-px) [reply tests/backup/backup_test.go:1414]: done

**Review by kshithijiyer-px — APPROVED** [review]
LGTM

**Review by lsrinivas-pure — APPROVED** [review]
LGTM

**Pull request merged**